content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+
+The task of formally specifying Pthreads presents an opportunity for rethinking its synchronization-oriented paradigm.
+The author identifies that synchronization is not always desirable, and that it requires a precise interleaving of expensive atomic operations:
+>The cost of atomic operations and memory barriers varies widely, but is often comparable to that of a hundred or more register-to-register instructions, even in the absence of a cache miss. For example, on some Pentium 4 processors, hardware instructions to atomically update a memory location require well over 100 processor cycles, and these can also double as one of the cheaper mechanisms for ensuring that a store operation becomes visible to other threads before a subsequent load. 
+
+Instead of synchronization, the author argues in favor of a paradigm that allows data races and relies on atomic operations by showing performance gains from such a paradigm shift. 
+>As a result of the high cost of these hardware instructions, and the even higher cost of the pthread primitives built on them, there are a small number of cases in which synchronization performance is critical, and more careful and direct use of the hardware primitives, together with less constrained use of shared variables, is essential. In some cases it may also be necessary to avoid deadlock issues inherent in lock-based programming[7], or desirable because a different parallel programming model is preferable for an application (cf. [32]).
+
+To demonstrate this claim, the author compares the performance of an implementation of the Sieve of Eratosthenes algorithm and a tracing garbage collector implemented under the synchronization and data raceful paradigm.
+
+Four implementations are evaluated when run with 1, 2, and 4 threads.  
+Two of these use synchronization: one using mutexes, and another using spinlocks.  
+These are compared to two implementations with data races: one that makes use of atomic memory operations, and one that unsafely uses ordinary memory operations.
+
+![alt text](images/sieve-bit-array.png "Sieve of Eratosthenes Performance")
+
+![alt text](images/gc.png "Tracing Garbage Collector Performance")
+
+The results indicate that there are applications for which allowing data races can be improve performance in a multi-processor, motivating the formal specification of programs with data races.
+
+## Formal Model for Data Races
+To allow data races, the language standard must provide a memory model for programs with data races.
+Java, as a safe language, provides such a model, so the author expressed optimism about the possibility of adapting its model for use in the C++ standard.
+
+Fast-forwarding to the present, however, we see that the problem of defining data races in C++ turned out to be intractable.
```

Comment:
In the context of the atomics discussion above, this might be a little misleading. The C++11 standardization effort _did_ succeed in giving semantics to lock-free programs that use atomic memory operations. So in that sense, it reclassified those programs which used to be "racy" as "race-free."

But there is still the matter of programs with "true" data races, i.e., concurrent normal/non-atomic accesses (to the same memory location, at least one of which is a write). C++11 still gives no semantics to these programs, and Java still tries to give semantics to them (but the semantics are extremely complicated).

Reply:
The authors don't make this distinction between "true" data races and racing atomics, which I find especially confusing in the context of DRF => SC: 
- Does DRF only exclude "true" data races (as opposed to "race conditions")?
- Should we call it TDRF ("true" data race freedom) => SC ?